lib/navigation/preprocess_encoders.hpp

```diff
@@ -1,27 +1,104 @@
 #pragma once
+#include "consts.hpp"
 
+#include <cmath>
+
+#include <algorithm>
 #include <array>
 #include <cstdint>
+#include <optional>
 
+#include "core/logger.hpp"
 #include "core/types.hpp"
 
 namespace hyped::navigation {
+
 class EncodersPreprocessor {
  public:
-  EncodersPreprocessor();
+  EncodersPreprocessor(core::ILogger &logger);
 
-  core::EncoderData processData(const core::EncoderData encoder_data);
+  /**
+   * @brief process encoder data to deal with outliers and unreliable sensors
+   *
+   * @param encoder_data
+   * @return cleaned and reliable encoder data
+   */
+  std::optional<core::EncoderData> processData(const core::EncoderData &encoder_data);
 
  private:
-  core::EncoderData detectOutliers(const core::EncoderData encoder_data);
+  struct Statistics {
+    core::Float median;
+    core::Float upper_bound;
+    core::Float lower_bound;
+  };
+
+  /**
+   * @brief calculate the upper and lower bounds as well as the median for
+   * the input data array
+   *
+   * @param encoder_data
+   * @return std::optional<Statistics>
+   */
+  std::optional<Statistics> getStatistics(const core::EncoderData &encoder_data) const;
+
+  /**
+   * @brief tidies up the received data by detecting the outliers or data received from faulty
+     sensors and replaces them with median value of the dataset; if there are too many
+     outliers or faulty sensors then this function fails
```

Comment:
be brief, people don't want to read a novel; keep the non-obvious parts and remove everything else

"detects new outliers and replaces values from faulty sensors with the median; fails if not enough reliable data is available"

if people want to know how exactly it works, they'll go read the implementation (which I would advise everyone to do anyways)